SCClassLibrary/Common/UnitTesting/UnitTest.sc

```diff
@@ -52,14 +52,18 @@ UnitTest {
 
 	// run all UnitTest subclasses
 	*runAll {
```

Comment:
Your solution above looks quite elegant, and would essentially make `run` and `runAll` methods equivalent for individual unit tests. Originally I thought it's not good to have two methods do the same thing, but maybe this is OK?

Reply:
They aren't exactly the same: `*run` doesn't look into subclasses.

I just found a concrete reason why `*runAll` might legitimately be made more forgiving.

```
UnitTest.allSubclasses.select { |class| class.subclasses.size > 0 }
-> [ class AbstractLPCBrutalTest ]

AbstractLPCBrutalTest.allSubclasses
-> [ class TestLexerBrutal, class TestLexerTargetedBrutal, class TestParserBrutal, class TestCompilerBrutal ]

AbstractLPCBrutalTest.findTestMethods
-> [  ]
```

So `AbstractLPCBrutalTest.run` will do nothing, but `AbstractLPCBrutalTest.runAll` will run the whole suite of compiler tests.

And, the current PR takes away the ability to run subsets of tests, organized by class hierarchy.

So it seems, in retrospect, that in fact it *is* meaningful for subclasses of UnitTest to implement `*runAll`, and an error disallowing that would be logically incorrect.

Reply:
thanks @jamshark70 for valuable insight! I changed the PR description accordingly.

Reply:
@jamshark70 AbstractLPCBrutalTest actually brings an interesting point - the problem is in a way "reverse" to the original one: `AbstractLPCBrutalTest.run` does not run any tests, and the output is 
```
RUNNING UNIT TEST 'AbstractLPCBrutalTest'

UNIT TESTS FOR 'AbstractLPCBrutalTest' COMPLETED
There were no failures
```
I think there may be a need to somehow query instance methods to see if there indeed are any tests, and then run them. `*run` should run only methods of the class, and I'm not sure what about `*runAll`, but maybe it should _only_ query subclasses to run their tests. In that case calling `*run` on a class that does not contain tests should probably return a warning, and calling `*runAll` on a class that does not contain subclasses should yield a warning as well... what do you think?

Reply:
I tried it in the new commit. `*run` queries test methods and prints warning if it doesn't find any. `*runAll` chocks if there are subclasses, _but it also adds test methods from current class_, in case there is a class that has test methods _and_ subclasses, and runs them. This makes `*runAll` also run test for a single class. Not sure if this is the best behavior, looking for feedback here...

Reply:
> I tried it in the new commit. *run queries test methods and prints warning if it doesn't find any.

too clever, besides in that case anyone can tell that the output doesn't run any tests.

> *runAll checks if there are subclasses, but it also adds test methods from current class, in case there is a class that has test methods and subclasses, and runs everything. This makes *runAll also run test for a single test class. Not sure if this is the best behavior, looking for feedback here...

this makes the concept of a test suite equivalent to inheritance, which is an unnecessary and incorrect coupling. `runAll` should _run all tests_, that's what the name says. if we want a concept of a test suite (which we definitely do), it shouldn't be tied to inheritance (especially in a language with single inheritance). i think your first instinct was totally correct -- only allow `runAll` to run on `UnitTest` for now.

seriously, i understand everyone's desire to fix the design of the unit testing framework, i agree it's not good, but we are just adding technical debt when we haphazardly implement features 'as needed' (i.e., because it seems good in the moment). fixing UnitTest in any meaningful way would be RFC-level work.